runtime/src/accounts_db.rs

```diff
@@ -713,26 +751,56 @@ impl AccountsDB {
         // Calculate store counts as if everything was purged
         // Then purge if we can
         let mut store_counts: HashMap<AppendVecId, (usize, HashSet<Pubkey>)> = HashMap::new();
-        let storage = self.storage.read().unwrap();
-        for (key, (account_infos, _ref_count)) in &purges {
-            for (slot, account_info) in account_infos {
-                let slot_storage = storage.0.get(&slot).unwrap();
-                let store = slot_storage.get(&account_info.store_id).unwrap();
+        for (key, (account_infos, ref_count)) in purges.iter_mut() {
+            if purged_account_slots.contains_key(&key) {
+                *ref_count = self
+                    .accounts_index
+                    .read()
+                    .unwrap()
+                    .ref_count_from_storage(&key);
+            }
+            account_infos.retain(|(slot, account_info)| {
+                let was_slot_purged = purged_account_slots
+                    .get(&key)
+                    .map(|slots_removed| slots_removed.contains(slot))
+                    .unwrap_or(false);
+                if was_slot_purged {
+                    // No need to look up the slot storage below if the entire
+                    // slot was purged
+                    return false;
+                }
+                // Check if this update in `slot` to the account with `key` was reclaimed earlier by
+                // `clean_old_rooted_accounts()`
+                let was_reclaimed = removed_accounts
+                    .get(&account_info.store_id)
+                    .map(|store_removed| store_removed.contains(&account_info.offset))
+                    .unwrap_or(false);
+                if was_reclaimed {
+                    return false;
+                }
                 if let Some(store_count) = store_counts.get_mut(&account_info.store_id) {
                     store_count.0 -= 1;
                     store_count.1.insert(*key);
                 } else {
                     let mut key_set = HashSet::new();
                     key_set.insert(*key);
-                    store_counts.insert(
-                        account_info.store_id,
-                        (store.count_and_status.read().unwrap().0 - 1, key_set),
+                    let count: usize = {
+                        let storage = self.storage.read().unwrap();
+                        let slot_storage = storage.0.get(&slot).unwrap();
+                        let store = slot_storage.get(&account_info.store_id).unwrap();
+                        let count = store.count_and_status.read().unwrap().0;
```

Comment:
how about defining a helper function at `AccountStorage`? I remember we have close one in tests.